### DISCUSSION TITLE
system.conf: allowed slot string format/names

### DISCUSSION THREAD

**Discussion by multiple1** [rauc/rauc#1430] [Q&A]
Hello!

Using an A/B system everyone uses A/B for bootname, but names the slots 0/1.
```
[slot.rootfs.0]
device=/dev/mmcblk0p2
type=ext4
bootname=A

[slot.rootfs.1]
device=/dev/mmcblk0p3
type=ext4
bootname=B
```
I think this is miss leading e.g. for rauc status mark-bad rootfs.1 i have to use 0/1

Is it fine to name the slots accordingly?
rauc status mark-bad roofs.B
```
[slot.rootfs.A]
device=/dev/mmcblk0p2
type=ext4
bootname=A

[slot.rootfs.B]
device=/dev/mmcblk0p3
type=ext4
bootname=B
```

**Comment by ejoerns** [comment]
@multiple1 Currently, there seem to be no assumptions in the parser code the slot `index` has to be an integer.

The main point of the index is to distinguish the different slot class members. Using `A` / `B` (/ `C`) should be sufficient for this case.

We currently document `<idx>` as the *slot number*, thus I am not fully sure if any currently developed feature relies on this but I would assume not.

Maybe in the future one could think of ordering implications based on the index, but as long as it is sortable, this might be sufficient.

@jluebbe Do you have any opinion on this?

**Reply by jluebbe** [reply]
In our tests and real-world deployments, we often uses "system0/1" for bootnames. :)

For numbers, the sorting is obvious. With strings we'd need to choose an order (perhaps just `strcmp`/ASCII order).

We should probably restrict slot class and index to alpha-numeric and warn if they contain something else.

**Comment by multiple1** [comment]
ok, so it should be fine.

I noticed that `rauc status` shows a different order if I change from 0/1 to A/B
system.conf:
rootfs.0
....
rootfs.1
=> rauc status shows 0 first 1 second

system.conf:
rootfs.A
....
rootfs.B
=> rauc status shows B first A second


system.conf
rootfs.A0
....
rootfs.B1
=> rauc status shows A first B second


For a/b system the order does not affect anything.
But how does the rauc install select the next slot if a c system is added?
e.g. B system is booted what is the next slot A or C? Depending on sorting?

**Reply by ejoerns** [reply]
Currently, the order is implementation-defined by the data types used. With the existing slot selection mechanism, a real A/B/C update is not possible. Some time ago, I made patches to always select the slot group with the oldest date in case there are multiple available. Will check if they are mainlinable if I find the time.

**Reply by ejoerns** [reply]
FTR: Meanwhile I submittest them as https://github.com/rauc/rauc/pull/1525
